yarn-project/sequencer-client/src/publisher/l1-publisher.ts

```diff
@@ -151,32 +101,110 @@ export type L1SubmitProofArgs = {
  */
-export class L1Publisher implements L2BlockReceiver {
+export class L1Publisher {
   private interruptibleSleep = new InterruptibleSleep();
   private sleepTimeMs: number;
   private interrupted = false;
   private metrics: L1PublisherMetrics;
   private log = createDebugLogger('aztec:sequencer:publisher');
 
-  constructor(private txSender: L1PublisherTxSender, client: TelemetryClient, config?: PublisherConfig) {
+  private availabilityOracleContract: GetContractReturnType<
+    typeof AvailabilityOracleAbi,
+    WalletClient<HttpTransport, chains.Chain, PrivateKeyAccount>
+  >;
+  private rollupContract: GetContractReturnType<
+    typeof RollupAbi,
+    WalletClient<HttpTransport, chains.Chain, PrivateKeyAccount>
+  >;
+  private publicClient: PublicClient<HttpTransport, chains.Chain>;
+  private account: PrivateKeyAccount;
+
+  constructor(config: TxSenderConfig & PublisherConfig, client: TelemetryClient) {
     this.sleepTimeMs = config?.l1PublishRetryIntervalMS ?? 60_000;
     this.metrics = new L1PublisherMetrics(client, 'L1Publisher');
+
+    const { l1RpcUrl: rpcUrl, l1ChainId: chainId, publisherPrivateKey, l1Contracts } = config;
+    const chain = createEthereumChain(rpcUrl, chainId);
+    this.account = privateKeyToAccount(publisherPrivateKey);
+    const walletClient = createWalletClient({
+      account: this.account,
+      chain: chain.chainInfo,
+      transport: http(chain.rpcUrl),
+    });
+
+    this.publicClient = createPublicClient({
+      chain: chain.chainInfo,
+      transport: http(chain.rpcUrl),
+    });
+
+    this.availabilityOracleContract = getContract({
+      address: getAddress(l1Contracts.availabilityOracleAddress.toString()),
+      abi: AvailabilityOracleAbi,
+      client: walletClient,
+    });
+    this.rollupContract = getContract({
+      address: getAddress(l1Contracts.rollupAddress.toString()),
+      abi: RollupAbi,
+      client: walletClient,
+    });
   }
 
   public async attest(archive: `0x${string}`): Promise<Attestation> {
-    return await this.txSender.attest(archive);
+    // @note  Something seems slightly off in viem, think it should be Hex instead of Hash
+    //        but as they both are just `0x${string}` it should be fine anyways.
+    const signature = await this.account.signMessage({ message: { raw: archive } });
+    const { r, s, v } = parseSignature(signature as `0x${string}`);
+
+    return {
+      isEmpty: false,
+      v: v ? Number(v) : 0,
+      r: r,
+      s: s,
+    };
+  }
+
+  public getSenderAddress(): Promise<EthAddress> {
+    return Promise.resolve(EthAddress.fromString(this.account.address));
   }
 
-  public async senderAddress(): Promise<EthAddress> {
-    return await this.txSender.getSenderAddress();
+  // Computes who will be the L2 proposer at the next Ethereum block
+  // Using next Ethereum block so we do NOT need to wait for it being mined before seeing the effect
+  // @note Assumes that all ethereum slots have blocks
```

Comment:
Is there an issue to fix this?

Reply:
What do you mean fix it? There is a tiny "diff" when slots are missed, but you don't really know if it will be missed or not before anything have even been broadcast so not sure how you would fix it.

The case where it is missed could likely handle the timing slightly better to compute the next potential block, but as long as there are plenty other issues it seems like a very tiny thing but can create an issue for it 🤷